### DISCUSSION TITLE
Caching 404 fetches in the data cache.

### DISCUSSION THREAD

**Discussion by nickjackson** [vercel/next.js#63239] [App Router]
Might there be appetite to provide an option in the [`NextFetchRequestConfig`](https://github.com/vercel/next.js/blob/3e6720c3647480d94b9dc01c4ed12988b06fa659/packages/next/types/global.d.ts#L53C11-L53C33) that allows for fetches that respond with 404 to be cached?

### Dynamic Routes
Most of the sites we build have dynamically driven pages, underpinned by some content in a CMS. We use dynamic routes to pass through the slug to the CMS which will return a 200 or 404 depending if there is content at that path or not, all pretty standard. We cache the 200, but if the response we get is a 404, no matter how many times that page is hit, it will always hit the origin and be slow. I appreciate that with this pattern of dynamic routes backed by a CMS, there is always going to be the ability to enumerate pages and hit the origin, but being able to cache the common cases helps reduce wasted round trips.

### Composing content
We make use of a pattern that allows our customers to build their own template overrides in their CMS, which can change the way a piece of content is rendered based on certain signals. For example, in the case of a PDP page, we would have a default template setup in our CMS such as `pdp/templates/default`, but the customer could override that for a particular brand, just by setting up a new piece of content at `pdp/templates/nike`

We then implement the above by doing two simultaneous fetches to retrieve the default template and one based on the brand of the current product. If the latter existed we would use that instead. If the override doesn't exist, then that will 404, hit the origin, and slow down the entire page.

### CMS based solutions
It is unlikely that the CMS would be willing to stop their pages from 404'ing, but they may offer alternative ways of fetching content that doesn't 404.

### `unstable_cache`
This is probably the only workable solution right now, but is unstable.

## Possible implementation
I'm really not tied to any implementation, but here are some ideas I have to expose this option.

```ts
interface NextFetchRequestConfig {
  revalidate?: number | false
  tags?: string[]
  includeNotFound?: boolean;
}
```

or perhaps

```ts
interface NextFetchRequestConfig {
  revalidate?: number | false
  tags?: string[]
  includeStatuses?: number[];
}
```

**Comment by dealer-solutions-gene** [comment]
I'm here because I too am building a content system, where my page routes are dynamic, and its the CMS's responsibility to respond with a valid 404 where it would be highly desirable to have the front end cache this successful transaction.

**Comment by alexhwoods** [comment]
We're hitting the same thing

**Comment by arvinpoddar** [comment]
We're in the same boat, our pages fetch data from an external source that often 404s. This results in page visits hitting the external source API, and we'd prefer not to rely on `unstable_cache` / `use cache` for compatibility reasons.

**Comment by nickjackson** [comment]
So I've had a few years to think about this, and `"use cache"` solves this somewhat, especially for the "composing content" use case I mentioned.

However, I don't believe we have a nice solution for "dynamic routes" though. I've found using `unstable_cache` isn't a workable for the simple reason that bots can visit pages never visited before extremely easily and increase the load and requests to your CMS. The 404's you can cache is a drop in the ocean compared to the routes you cannot predict. 

What I think could be handy is something like `generateAllowedParams`, similar to `generateStaticParams`, but used for gating instead of prerendering. 

```ts
// app/blog/[slug]/page.tsx

export async function generateAllowedParams(): Promise<{ slug: string }[]> {
  const res = await fetch('https://cms.example.com/api/slugs?fields=slug&status=published')
  const posts: { slug: string }[] = await res.json()
  return posts
}

export const allowedParamsConfig = {
  revalidate: 300,    
  tags: ['cms-slugs'],
}
```

I think this could be built today with something like the below, but I would imagine building into next.js means the situation can be optimized for

```ts
// app/blog/[slug]/layout.tsx
import { notFound } from 'next/navigation'
import { cacheTag, cacheLife } from 'next/cache'

async function getAllowedSlugs(): Promise<Set<string>> {
  'use cache'
  cacheTag('cms-slugs')
  cacheLife({ revalidate: 300 })

  const res = await fetch('https://cms.example.com/api/slugs?fields=slug&status=published')
  const posts: { slug: string }[] = await res.json()
  return new Set(posts.map(p => p.slug))
}

export default async function BlogSlugLayout({
  children,
  params,
}: {
  children: React.ReactNode
  params: { slug: string }
}) {
  const allowed = await getAllowedSlugs()

  if (!allowed.has(params.slug)) {
    notFound()
  }

  return children
}
```
